### PR TITLE
Extend backoff supervision with custom message handling #29082

### PR DESCRIPTION
**Pull request by nvollmar**
Refs #29082
Allows for more flexible replying while supervised actor is stopped.

### CODE REVIEW

File: akka-actor-tests/src/test/scala/akka/pattern/BackoffSupervisorSpec.scala [hunk 1]
```diff
@@ -203,11 +202,36 @@ class BackoffSupervisorSpec extends AkkaSpec with ImplicitSender with Eventually
       }
     }
 
+    "reply to sender if withHandlerWhileStopped is specified" in {
+      filterException[TestException] {
+        val supervisor = create(
+          BackoffOpts
+              .onFailure(Child.props(testActor), "c1", 100.seconds, 300.seconds, 0.2)
+              .withMaxNrOfRetries(-1)
+              .withHandlerWhileStopped({case (msg, sender) => sender ! TestReply(msg)}))
```
Comment (johanandren) [inline akka-actor-tests/src/test/scala/akka/pattern/BackoffSupervisorSpec.scala:211]: While I understand that this gives the ultimate flexibility, it looks a bit dangerous to me because easy to accidentally/unsafely close over the surrounding scope, it essentially is giving an actor (the supervisor) a receive block from the outside.

Could it perhaps be a stand in actor props instead?

Comment (nvollmar) [reply akka-actor-tests/src/test/scala/akka/pattern/BackoffSupervisorSpec.scala:211]: You mean props of a custom actor receiving the message and replying while the supervised is stopped?

Comment (johanandren) [reply akka-actor-tests/src/test/scala/akka/pattern/BackoffSupervisorSpec.scala:211]: Yep, not sure it is the greatest of ideas but it would be safer.

Comment (nvollmar) [reply akka-actor-tests/src/test/scala/akka/pattern/BackoffSupervisorSpec.scala:211]: I like the idea, it does not open the API for easy to make mistakes while providing the flexibility we need :)

Comment (nvollmar) [reply akka-actor-tests/src/test/scala/akka/pattern/BackoffSupervisorSpec.scala]: I've pushed a new commit with the proposed change

File: akka-actor/src/main/scala/akka/pattern/HandleBackoff.scala [hunk 2]
```diff
@@ -32,6 +33,15 @@ import akka.annotation.InternalApi
     child = Some(context.watch(context.actorOf(childProps, childName)))
   }
 
+  protected def getOrCreateHandler(p: Props): ActorRef = handler match {
+    case Some(h) =>
+      h
+    case None =>
+      val h = context.actorOf(p)
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/HandleBackoff.scala:40]: so the handler is started on demand, but never stopped?

I wonder if an alternative would be to be able to pass in an `ActorRef` instead of a `Props`. Kind of a way to replace the `deadLetters` ActorRef?

Comment (nvollmar) [reply akka-actor/src/main/scala/akka/pattern/HandleBackoff.scala:40]: It could be stopped once the child is restarted again.

Just accepting an actorRef would be the most light-weight solution and provide the same functionality.

Comment (nvollmar) [reply akka-actor/src/main/scala/akka/pattern/HandleBackoff.scala:40]: Pro/Cons:

With `Props` we can start handling actor only when we need it (assuming it's an exceptional case) but each supervisor would start its own.
With `ActorRef` the API user can share the handling actor across all supervised actors if desired, but would have to be eager instantiated.

Comment (nvollmar) [reply akka-actor/src/main/scala/akka/pattern/HandleBackoff.scala]: I've pushed another update just using `ActorRef`. Adds the least complexity for special cases.

File: akka-actor/src/main/scala/akka/pattern/internal/BackoffOnStopSupervisor.scala [hunk 3]
```diff
@@ -88,9 +88,9 @@ import scala.concurrent.duration.FiniteDuration
         case Some(fsm) if fsm(msg) => context.stop(self)
         case _ =>
           handlerWhileStopped match {
-            case None           => context.system.deadLetters.forward(msg)
             case Some(Left(r))  => sender() ! r
-            case Some(Right(p)) => getOrCreateHandler(p).forward(msg)
+            case Some(Right(h)) => h.forward(msg)
+            case None           => context.system.deadLetters.forward(msg)
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/internal/BackoffOnStopSupervisor.scala:93]: could we even remove the Option for `handlerWhileStopped`? `Either[Any, ActorRef]` and populate the Right with system.deadLetters if the custom ActorRef is not given?

Comment (nvollmar) [reply akka-actor/src/main/scala/akka/pattern/internal/BackoffOnStopSupervisor.scala:93]: quickly looked into it, but this requires access to the system/context which is only available here.

Comment (nvollmar) [reply akka-actor/src/main/scala/akka/pattern/internal/BackoffOnStopSupervisor.scala:93]: Would be a little simpler, but requires access to context/system (which would be a breaking change to the API to have it there).

Comment (patriknw) [reply akka-actor/src/main/scala/akka/pattern/internal/BackoffOnStopSupervisor.scala:93]: ok, np

File: akka-actor-tests/src/test/scala/akka/pattern/BackoffSupervisorSpec.scala [hunk 4]
```diff
@@ -203,11 +201,43 @@ class BackoffSupervisorSpec extends AkkaSpec with ImplicitSender with Eventually
       }
     }
 
+    "reply to sendt akker if withHandlerWhileStopped is specified" in {
```
Comment (johanandren) [inline akka-actor-tests/src/test/scala/akka/pattern/BackoffSupervisorSpec.scala:204]: Hard to parse test name, what about "use alternative actor while stopped"?

File: akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala [hunk 5]
```diff
@@ -299,6 +298,15 @@ private[akka] sealed trait ExtendedBackoffOptions[T <: ExtendedBackoffOptions[T]
...
   def withReplyWhileStopped(replyWhileStopped: Any): T
 
+  /**
+   * Returns a new BackoffOptions with a custom handler for messages that the supervisor receives while its child is stopped.
+   * By default, a message received while the child is stopped is forwarded to `deadLetters`.
+   * Essentially, this handler replaces `deadLetters` allowing to implement custom handling instead of a static reply.
+   *
+   * @param handler PartialFunction of the received message and sender
+   */
+  def withHandlerWhileStopped(handler: ActorRef): T
```
Comment (johanandren) [inline akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:308]: One thing with going for an actorRef instead of props is that it will be a shared actor, for example when used with sharding you will have to spawn a single alternative actor for an entire node. Not sure if that is a problem  or if that is ok.

Nice to not have to manage lifecycle of the stopped handler though. Hmm.

File: akka-actor/src/main/scala/akka/pattern/internal/BackoffOnRestartSupervisor.scala [hunk 6]
```diff
@@ -94,9 +94,10 @@ import akka.pattern.{ BackoffReset, BackoffSupervisor, HandleBackoff }
     case Some(c) =>
       c.forward(msg)
     case None =>
-      replyWhileStopped match {
-        case None    => context.system.deadLetters.forward(msg)
-        case Some(r) => sender() ! r
+      handlerWhileStopped match {
+        case Some(Left(r))  => sender() ! r
+        case Some(Right(h)) => h.forward(msg)
+        case None           => context.system.deadLetters.forward(msg)
```
Comment (johanandren) [inline akka-actor/src/main/scala/akka/pattern/internal/BackoffOnRestartSupervisor.scala:100]: Not super important since internal, but perhaps encode this in an internal sealed ADT instead of nesting Option and Either, given that it basically is three possible states now?

**Review by johanandren — COMMENTED** [review]
I think this is a better way to go. Added some smaller comments.

File: akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala (hunk 5 above)
Comment (aluveitie) [reply akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:308]: In the end it's up to the user. You can have one actor for a node (which in most cases would probably suffice), or spawn one per supervisor as sibling if you need that.

Comment (nvollmar) [reply akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:308]: In the end it's up to the user. You can have one actor for a node (which in most cases would probably suffice), or spawn one per supervisor as sibling if you need that.

File: akka-actor/src/main/scala/akka/pattern/internal/BackoffOnRestartSupervisor.scala (hunk 6 above)
Comment (nvollmar) [reply akka-actor/src/main/scala/akka/pattern/internal/BackoffOnRestartSupervisor.scala]: Yep, pushed an update with a sealed ADT, makes it more readable 👍

File: akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala (hunk 5 above)
Comment (johanandren) [reply akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:308]: Ah, you could indeed just put another actor in between. I think that is good enough.

File: akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala [hunk 7]
```diff
@@ -431,10 +431,17 @@ private final case class BackoffOnFailureOptionsImpl[T](
...
+@InternalApi
 private[akka] sealed trait BackoffReset
 private[akka] case object ManualReset extends BackoffReset
 private[akka] final case class AutoReset(resetBackoff: FiniteDuration) extends BackoffReset
+
+@InternalApi
+private[akka] sealed trait HandlingWhileStopped
+private[akka] case object ForwardDeathLetters extends HandlingWhileStopped
+private[akka] case class ForwardTo(handler: ActorRef) extends HandlingWhileStopped
+private[akka] case class ReplyWith(msg: Any) extends HandlingWhileStopped
```
Comment (johanandren) [inline akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:447]: Good, much more obvious! 👍

**Review by johanandren — COMMENTED** [review]
LGTM, but something with the bincomp excludes now.

**Comment by nvollmar** [comment]
@johanandren Forgot to adapt the excludes - pushed a fix

**Review by johanandren — APPROVED** [review]

**Comment by aluveitie** [comment]
@patriknw Would be nice if we could get that into the next release

**Comment by nvollmar** [comment]
@patriknw Would be nice if we could get that into the next release

**Review by patriknw — APPROVED** [review]
LGTM, thanks for the  ping
